src/instructlab/eval/mt_bench_common.py

```diff
@@ -270,17 +273,16 @@ def chat_completion_openai(
                 temperature=temperature,
                 max_tokens=max_tokens,
             )
-            output = response.choices[0].message.content
-            break
-        except openai.OpenAIError as e:
-            if i == API_MAX_RETRY - 1:
-                # Print error on last try
-                print(type(e), e)
-            else:
-                logger.debug(e)
-            time.sleep(API_RETRY_SLEEP)
-
-    return output
+            return response.choices[0].message.content
+        except openai.APITimeoutError as ex:
+            logger.debug(ex)
+        except openai.APIConnectionError as ex:
```

Comment:
One error condition we should be able to create is overloading llama-cpp cpu serving with max-workers.  That tends to create a service unavailable condition.  Would be good to know where that hits in here.

Also, when calling against the openai api, do we know for sure this isn't a condition we hit during normal use?  They were retrying up to 16 times previously but I don't know the common errors returned.

Reply:
I honestly don't know how often 40x return from API. I will rely on your judgement. (Which I think suggests it's a common scenario.)

---

My primary goal is to make the code that calls this function to fail (to report back to CI that it's broken). Without it, e2e runs claim success when in reality evaluation crashed. I think CI status should be indicative of a failure - one should not have to drill down into logs to double-check a job actually worked.

But: I think this can be achieved while leaving the retry logic for `APIConnectionError` in place. Just need to `raise` instead of `return output`. This will allow us to retry for some time but still fail if an error happened.

What do you think?

Reply:
I am a +1 to failing fast for any conditions which are known to be invalid.  No model being served is one of those.  As long as we can differentiate that case, let's do it.  The two conditions we know of that we need to make sure still work are:

- Getting 400s because of max context length issues.  You would get this using granite as the judge model for example.
- llama-ccp cpu serving being overloaded and returning service unavailable.

Neither of those conditions should fail fast and we just need to make sure they don't hit this code block.  I would assume the 400 error does not.  But I don't know about the service unavailable.

For the context length issues additional improvements might be:
- Don't retry when you hit the issue the first time but record the error and move on
- Default to failing fast with this condition and allow for an environment variable to override

The case I am not comfortable with yet is failing after the retries run out vs recording the error in the error rate.  There is more discussion on this in response to other comments.